### PR TITLE
AbstractReversibleSolver + ReversibleAdjoint

### PR DESCRIPTION
**Pull request by sammccallum**
Implements `AbstractReversibleSolver` base class and `ReversibleAdjoint` for reversible back propagation.

This updates `SemiImplicitEuler`, `LeapfrogMidpoint` and `ReversibleHeun` to subclass `AbstractReversibleSolver`.

### Implementation
`AbstractReversibleSolver` subclasses `AbstractSolver` and adds a `backward_step` method:
```python
@abc.abstractmethod
def backward_step(
    self,
    terms: PyTree[AbstractTerm],
    t0: RealScalarLike,
    t1: RealScalarLike,
    y1: Y,
    args: Args,
    solver_state: _SolverState,
    made_jump: BoolScalarLike,
) -> tuple[Y, DenseInfo, _SolverState]:
```
This method should reconstruct `y0`, `solver_state` at `t0` from `y1`, `solver_state` at `t1`. See the aforementioned solvers for examples.

When backpropagating, `ReversibleAdjoint` uses this `backward_step` to reconstruct state. We then take a `vjp` through a local forward step and accumulate gradients.

`ReversibleAdjoint` now also pulls back gradients from any interpolated values, so we can use `SaveAt(ts=...)`!

We allow arbitrary `solver_state` (provided it can be reconstructed reversibly) and calculate gradients w.r.t. `solver_state`. Finally, we pull back these gradients onto `y0`, `args`, `terms` using the `solver.init` method.

### CODE REVIEW

**Comment by sammccallum** [comment]
Implements #541 and #584

**Comment by patrick-kidger** [comment]
This looks very clean - I have almost no notes!

I think we should make sure that this also works for creating reversible RK methods before merging, but fundamentally I think I'm happy with this.

File: diffrax/_adjoint.py [hunk 1]
```diff
@@ -908,3 +907,278 @@ def loop(
...
+                    "However `ReversibleAdjoint` currently only supports Stratonovich "
+                    "SDEs."
+                )
+            elif not isinstance(solver, AbstractStratonovichSolver):
+                warnings.warn(
+                    f"{solver.__class__.__name__} is not marked as converging to "
+                    "either the Itô or the Stratonovich solution. Note that "
+                    "`ReversibleAdjoint` will only produce the correct solution for "
+                    "Stratonovich SDEs."
+                )
```
[the remark this answers is not in the document]
Comment (sammccallum) [reply diffrax/_adjoint.py:1160]: Yes I think you're right. The solver converging to Statonovich is a sufficient but not a necessary condition.

Having said that, you could imagine the forward step converging to the Itô solution while the backwards step converges to the backwards Itô solution (e.g. this is the case for James' reversible method applied to Euler-Maruyama).

But, I think should this be handled by the `AbstractReversibleSolver` and not `ReversibleAdjoint`. I can add something to the docstring in ReversibleSolver (also I guess if the user is writing reversible SDE solvers they're probably aware of this).

File: diffrax/_adjoint.py [hunk 2]
```diff
@@ -908,3 +907,278 @@ def loop(
...
+            grad_y = (ω(grad_ys)[saveat_ts_index]).ω
+            _, interp_vjp = eqx.filter_vjp(interpolate, t, t0, t1, dense_info)
+            interp_grads = interp_vjp(grad_y)
+            grad_dense_info = eqx.apply_updates(grad_dense_info, interp_grads[3])
+            saveat_ts_index = saveat_ts_index - 1
+            return saveat_ts_index, grad_dense_info
+
+        grad_dense_info = jtu.tree_map(jnp.zeros_like, dense_info)
+        inner_state = (saveat_ts_index, grad_dense_info)
+        inner_state = eqxi.while_loop(_cond_fun, _body_fun, inner_state, kind="lax")
```
[the remark this answers is not in the document]
Comment (sammccallum) [reply diffrax/_adjoint.py]: I believe the only common logic is the use of the `interpolator.evaluate` method which is already factored out (plus the condition that we only use `dense_info` within `t0` and `t1`). 

The `cond_fun` is different because we are running backwards in time rather than forwards and the `body_fun` is different because we are reading from an array rather than setting.

But, I'm happy to try to abstract the core interpolation assumptions if you think that would be valuable.

File: diffrax/_solver/base.py [hunk 3]
```diff
@@ -348,3 +348,60 @@ def func(
...
+        - `args`: Any extra arguments passed to the vector field.
+        - `solver_state`: Any evolving state for the solver itself, at `t1`.
+        - `made_jump`: Whether there was a discontinuity in the vector field at `t1`.
+            Some solvers (notably FSAL Runge--Kutta solvers) usually assume that there
+            are no jumps and for efficiency re-use information between steps; this
+            indicates that a jump has just occurred and this assumption is not true.
+
+        **Returns:**
+
+        A tuple of several objects:
```
[the remark this answers is not in the document]
Comment (sammccallum) [reply diffrax/_solver/base.py]: The only idea here was to match the function signature of the forward `step`. But, I'm struggling to think of an example where you would require an error estimate on the `backward_step` (as we want to reconstruct our forward solve exactly).

For now, I'll remove the error estimate and results from the `backward_step` return.

**Comment by sammccallum** [comment]
Thanks very much for the review, Patrick. It's an honour for your code to be called "clean" from the sensei himself!

This implementation works for creating reversible RK solvers, but I'd like to work a bit more on that side of things before we look to merge. I can follow up in an email about the details.

I believe I have rebased on top of `dev` but If I've done something funky here please let me know!

**Comment by patrick-kidger** [comment]
This isn't intentionally closed -- this is just bad GitHub UX when a target branch (`dev`) is merged.

File: diffrax/_solver/base.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (patrick-kidger) [reply diffrax/_solver/base.py]: Sounds good!

File: diffrax/_adjoint.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (patrick-kidger) [reply diffrax/_adjoint.py]: Nope, fair enough to leave this as-is if you think that's the only thing in common.

File: diffrax/_adjoint.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (patrick-kidger) [reply diffrax/_adjoint.py]: So my point here is that reversible methods should obtain exactly the same values (up to floating point error) as directly backprop'ing through the solver.

Correspondingly concerns about Ito-vs-Strat-etc don't even come into it -- as long as you were comfortable backprop'ing through the internal numerics (which I think is always fine) then this should be fine too.

I'm not sure if we're agreeing with each other or not :D
